[PM-13128] Enable Breadcrumb Policies

jonashendrickx opened the pull request:
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-13128

## 📔 Objective

- Managed organizations by providers shouldn't be able to see or use the badges.
- Policies page is to be shown when subscribed to a Teams or Teams Starter plan.
- When conditions are met mentioned in bullet point number 1, we also show relevant `Upgrade` badges to upsell the `Enterprise` product.
- `org-permissions.guard.ts` allows for using async calls with injected services in the lambda expression or callback. This could potentially allow greater flexibility in the future.

## 📸 Screenshots


https://github.com/user-attachments/assets/7e84f279-fddf-48e7-b14b-b20d51c86945



## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

amorask-bitwarden commented on libs/common/src/billing/abstractions/organizations/organization-upselling.service.abstraction.ts:
```diff
@@ -0,0 +1,9 @@
+   * Determines if upselling policies is enabled for the organizations meeting certain criteria.
+   * @param organization
+   */
+  abstract isUpsellingPoliciesEnabled(organization: Organization): Promise<boolean>;
```
❓ What else is going to go in this service? I'd prefer to move this into the `OrganizationBillingService` so we're not managing services with 1 method and a very specific purpose.

amorask-bitwarden commented on libs/common/src/billing/services/organization/organization-upselling.service.ts:
```diff
@@ -0,0 +1,26 @@
+export class OrganizationUpsellingService implements OrganizationUpsellingServiceAbstraction {
+  constructor(private readonly configService: ConfigService) {}
+
+  async isUpsellingPoliciesEnabled(organization: Organization): Promise<boolean> {
```
🧪 Can we add some unit tests for this functionality?

amorask-bitwarden commented on apps/web/src/app/admin-console/organizations/guards/advanced-org-permissions.guard.ts:
```diff
@@ -0,0 +1,117 @@
+ * 2. If the logged in user does have the required permissions navigation
+ *    proceeds as expected.
+ */
+export function advancedOrganizationPermissionsGuard(
```
🤔 I like the idea, but this is an entirely new copy of an existing guard that will have to maintain the same logic as the prior one, albeit with the addition of a `services` property. Any reason the `organizationPermissionsGuard` couldn't be expanded to include optional services?

amorask-bitwarden commented on apps/web/src/app/admin-console/organizations/layouts/organization-layout.component.ts:
```diff
@@ -140,6 +143,14 @@ export class OrganizationLayoutComponent implements OnInit {
       ? "claimedDomains"
       : "domainVerification";
+
+    this.canShowPoliciesTab$ = this.organization$.pipe(
```
👍 Nice

amorask-bitwarden commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.ts:
```diff
@@ -114,4 +130,18 @@ export class PolicyEditComponent implements AfterViewInit {
+
+    await lastValueFrom(reference.closed);
+
+    this.dialogRef.close();
```
❓ Do we need to reload anything here or somehow move the user to another step?

jonashendrickx replied:
I've added a check for the dialog result of the change payment dialog and correctly close the policy edit dialog to notify the parent to refresh the data being displayed.

amorask-bitwarden reviewed (commented):
Just a couple thoughts here.

amorask-bitwarden commented on apps/web/src/app/admin-console/organizations/policies/reset-password.component.html:
```diff
@@ -2,7 +2,7 @@
 </bit-callout>
 
-<bit-callout type="success" [title]="'prerequisite' | i18n" icon="bwi-lightbulb">
+<bit-callout type="info" [title]="'prerequisite' | i18n" icon="bwi-lightbulb">
```
❓ Was this intended?

jonashendrickx replied:
Yes, it was mentioned in the Figma designs.

amorask-bitwarden reviewed (approved):
Thank you for making those changes, Billing stuff looks good to me. Just left a non-blocking question.

BTreston commented on apps/web/src/app/admin-console/organizations/policies/policies.component.ts:
```diff
@@ -37,27 +40,23 @@ export class PoliciesComponent implements OnInit {
-        this.organizationService
-          .organizations$(userId)
-          .pipe(getOrganizationById(this.organizationId)),
+        this.organizationService.getOrganizationById$(this.organizationId),
```
We should fix this to use the existing function in the organizaiton service:
```suggestion
this.organization$ = this.accountService.activeAccount$.pipe(
        getUserId,
        switchMap(userId => this.organizationService.organizations$(userId)),
        getOrganizationById(this.organizationId)
      );
```

BTreston commented on apps/web/src/app/admin-console/organizations/policies/policies.component.html:
we can change `organization` here to use the new `organization$` observable and the `async` pipe

BTreston commented on libs/common/src/admin-console/services/organization/default-organization.service.ts:
```diff
@@ -81,6 +84,12 @@ export class DefaultOrganizationService implements InternalOrganizationServiceAb
     return this.organizationState(userId).state$.pipe(this.mapOrganizationRecordToArray());
   }
 
+  getOrganizationById$(organizationId: string): Observable<Organization> {
```
we should delete this and use the existing methods per comments in `policy.component.ts`. This adds back usage of the active user state, which we want to avoid. For getting user ID, recomended approach is `accountService.activeAccount$.pipe(getUserId);`

BTreston commented on libs/common/src/billing/abstractions/organization-billing.service.ts:
```diff
@@ -59,4 +61,10 @@ export abstract class OrganizationBillingServiceAbstraction {
+   * Determines if upselling policies is enabled for the organizations meeting certain criteria.
+   * @param organization
+   */
+  abstract isUpsellingPoliciesEnabled(organization: Organization): Promise<boolean>;
```
Personally, I'd suggest updating this signature to return an `Observable<boolean>` and using rxjs for the implementation, but not a hill I'll die on. This would also make it integrate nicer with existing rxjs in the `OrganizationService`.

BTreston commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.ts:
```diff
@@ -92,6 +105,12 @@ export class PolicyEditComponent implements AfterViewInit {
+    );
+    this.isUpsellingEnabled = await this.organizationBillingService.isUpsellingPoliciesEnabled(
+      this.organization,
+    );
```
following the comment on updating `isUpSellingPoliciesEnabled` to return an `Observable<boolean>`:
```suggestion
this.organization$ = this.accountService.activeAccount$.pipe(
      getUserId,
      switchMap(userId => this.organizationService.organizations$(userId)),
      getOrganizationById(this.data.organizationId)
    );

    this.isUpsellingEnabled$ = this.organization$.pipe(
      switchMap(organization => this.organizationBillingService.isUpsellingPoliciesEnabled$(organization))
    );

```

BTreston commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.ts:
```diff
@@ -43,22 +50,28 @@ export class PolicyEditComponent implements AfterViewInit {
     enabled: [this.enabled],
   });
+  private organization: Organization;
+  protected isUpsellingEnabled: boolean;
```
`isUpsellingEnabled$: Observable<boolean>;`

BTreston commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.html:
```diff
@@ -1,5 +1,10 @@
 <form [formGroup]="formGroup" [bitSubmit]="submit">
   <bit-dialog [loading]="loading" [title]="'editPolicy' | i18n" [subtitle]="policy.name | i18n">
+    <ng-container bitDialogTitle>
+      <span bitBadge variant="primary" *ngIf="isUpsellingEnabled" (click)="changePlan()">
```
```suggestion
@let organization = (organization$ | async);
      <span bitBadge variant="primary" *ngIf="(isUpsellingEnabled$ | async)" (click)="changePlan(organization)">
```

BTreston commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.ts:
```diff
@@ -114,4 +133,22 @@ export class PolicyEditComponent implements AfterViewInit {
     return dialogService.open<PolicyEditDialogResult>(PolicyEditComponent, config);
   };
+
+  protected async changePlan() {
```
add the organization as a param so it can be passed in via the template and the async pipe.

BTreston reviewed (changes requested):
Got some suggestions, looking good so far, just a few things I think we can tweak

BTreston commented on apps/web/src/app/admin-console/organizations/layouts/organization-layout.component.ts:
```diff
@@ -140,6 +143,14 @@ export class OrganizationLayoutComponent implements OnInit {
       : "domainVerification";
+
+    this.canShowPoliciesTab$ = this.organization$.pipe(
+      switchMap((organization) =>
+        this.organizationBillingService
+          .isUpsellingPoliciesEnabled$(organization)
+          .pipe(map((isUpsellingEnabled) => isUpsellingEnabled || organization.canManagePolicies)),
```
⛏️ best practice for rxjs is to avoid nesting pipes. You should be able to do:
```suggestion
switchMap((organization) => this.organizationBillingService.isUpsellingPoliciesEnabled$(organization)),
map((isUpsellingEnabled) => isUpsellingEnabled || organization.canManagePolicies)
```

BTreston reviewed (changes requested):
Looks good, thank you for making the requested changes. Just a small ⛏️ and some test failures to look at

BTreston commented on libs/common/src/billing/services/organization-billing.service.ts:
```diff
@@ -220,4 +226,29 @@ export class OrganizationBillingService implements OrganizationBillingServiceAbs
     await this.billingApiService.restartSubscription(organizationId, request);
   }
+
+  isUpsellingPoliciesEnabled$(organization: Organization): Observable<boolean> {
```
This should return a `boolean` after the feature flag is gone per previous comment. 🎨 you can use the `map` operator here to avoid having to constantly return `of(x)` and simply return `x`

BTreston commented on apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts:
```diff
@@ -42,7 +47,10 @@ import { ToastService } from "@bitwarden/components";
  */
 export function organizationPermissionsGuard(
-  permissionsCallback?: (organization: Organization) => boolean,
+  permissionsCallback?: (
+    organization: Organization,
+    services: InjectedOrganizationPermissionServices,
+  ) => boolean | Promise<boolean>,
```
❓ For now this is okay, but we should refactor this to just return a boolean after the feature flag is gone. Can we add a todo message here to refactor this later?

eliykat replied:
I'm actually OK with leaving this because this requirement does come up from time to time - usually around permission changes behind a feature flag.

BTreston reviewed (changes requested):
Just two small notes, otherwise looks good

eliykat commented on apps/web/src/app/admin-console/organizations/policies/policies.component.ts:
```diff
@@ -109,4 +120,25 @@ export class PoliciesComponent implements OnInit {
+  protected async changePlan(organization: Organization) {
+    const reference = openChangePlanDialog(this.dialogService, {
+      data: {
+        organizationId: this.organizationId,
+        subscription: null,
+        productTierType: organization.productTierType,
```
Please use a consistent source of data. Either use the class state (`this.organizationId`) or the argument (`organization`) but not both.

eliykat commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.ts:
```diff
@@ -114,4 +143,22 @@ export class PolicyEditComponent implements AfterViewInit {
+
+  protected async changePlan(organization: Organization) {
+    const reference = openChangePlanDialog(this.dialogService, {
+      data: {
+        organizationId: this.data.organizationId,
+        subscription: null,
+        productTierType: organization.productTierType,
+      },
```
Same here

eliykat commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.ts:
```diff
@@ -114,4 +143,22 @@ export class PolicyEditComponent implements AfterViewInit {
+
+    const dialogResult = await lastValueFrom(reference.closed);
+
+    if (dialogResult === ChangePlanDialogResultType.Submitted) {
+      this.dialogRef.close(PolicyEditDialogResult.Saved);
```
Is this accurate? `PolicyEditDialogResult.Saved` indicates that the user has successfully edited a policy, not that they've upgraded their plan.

jonashendrickx replied:
I've added `Changed`, for the lack of a better idea of what to name it.

eliykat commented on apps/web/src/app/admin-console/organizations/settings/organization-settings-routing.module.ts:
```diff
@@ -41,7 +45,16 @@ const routes: Routes = [
-        canActivate: [organizationPermissionsGuard((org) => org.canManagePolicies)],
+        canActivate: [
+          organizationPermissionsGuard(
+            async (o: Organization, services: InjectedOrganizationPermissionServices) => {
+              const isUpsellingEnabled = await firstValueFrom(
+                services.organizationBillingService.isUpsellingPoliciesEnabled$(o),
+              );
+              return o.canManagePolicies || isUpsellingEnabled;
```
Suggestion: if the guard accepts a function that returns an observable (rather than a promise), you can just pipe your observable rather than using `firstValueFrom`. Given how we use rxjs that is probably a more common (and better) pattern.

eliykat commented on libs/common/src/billing/abstractions/organization-billing.service.ts:
```diff
@@ -59,4 +62,10 @@ export abstract class OrganizationBillingServiceAbstraction {
+   * Determines if upselling policies is enabled for the organizations meeting certain criteria.
+   * @param organization
+   */
+  abstract isUpsellingPoliciesEnabled$(organization: Organization): Observable<boolean>;
```
Suggestion: use consistent language - event logs called this "breadcrumbing", this calls it "upselling".

eliykat replied:
Nice work encapsulating the domain specific logic in your service rather than in the components 🎉

eliykat reviewed (changes requested):
I'm not sure if you meant to request another review as most feedback is unaddressed. Let me know if you want to discuss.

eliykat commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.ts:
```diff
@@ -30,6 +42,7 @@ export type PolicyEditDialogData = {
 
 export enum PolicyEditDialogResult {
   Saved = "saved",
+  Changed = "changed",
```
This can be specific to the action it represents: `UpgradedPlan`.

eliykat reviewed (changes requested):
As mentioned in Slack, I have some changes in #14041 which will enable this PR. That can be merged into this PR or into `main` depending on your preference.

Design feedback:

1. The Upgrade badge next to the Policies heading seems misaligned - it's missing spacing between it and the heading next to it.

![Screenshot 2025-03-28 at 11 12 38 AM](https://github.com/user-attachments/assets/a57c7d3c-ce36-4c05-9037-029add10600b)

2. Hovering over the Upgrade / Enterprise badges shows a text cursor rather than a button cursor. Looking at the HTML again, they're `<span>` elements when they should be button elements. (Better semantic HTML and accessibility as well.)

3. The upgrade modal opens up over the top of the edit policy modal. It's a question for your designer but I'm guessing this is probably not what we want. Maybe it would be better to close the policy modal first, then open the upgrade modal. You could use the Result type to handle this.

![Screenshot 2025-03-28 at 11 11 21 AM](https://github.com/user-attachments/assets/ec7a6361-9cd9-4d58-a2ec-1f63abeb591b)

3. The designs show that when breadcrumbing applies, the edit policy modal's primary button should read "Upgrade" instead of "Submit". That change is missing from this PR. I can still try to submit a policy change and I get an error message back from the server.

jonashendrickx commented:
> As mentioned in Slack, I have some changes in #14041 which will enable this PR. That can be merged into this PR or into `main` depending on your preference.
> 
> Design feedback:
> 
> 1. The Upgrade badge next to the Policies heading seems misaligned - it's missing spacing between it and the heading next to it.
> 
> ![Screenshot 2025-03-28 at 11 12 38 AM](https://private-user-images.githubusercontent.com/31796059/427803348-a57c7d3c-ce36-4c05-9037-029add10600b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0MTA2ODksIm5iZiI6MTc0MzQxMDM4OSwicGF0aCI6Ii8zMTc5NjA1OS80Mjc4MDMzNDgtYTU3YzdkM2MtY2UzNi00YzA1LTkwMzctMDI5YWRkMTA2MDBiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDA4Mzk0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllNDVmOWZmNGZhNzJjYmM0NWI2NjE5MGJlNTlmOTlmZWI0N2M3NTI5MjliZDliZmNmOTViMjU0ZGU3NzAwODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vhYN3StZGmVXz7hCC0fH4Fg4n7k_p4R-NMzuEaIiA0g)
> 
> 2. Hovering over the Upgrade / Enterprise badges shows a text cursor rather than a button cursor. Looking at the HTML again, they're `<span>` elements when they should be button elements. (Better semantic HTML and accessibility as well.)
> 3. The upgrade modal opens up over the top of the edit policy modal. It's a question for your designer but I'm guessing this is probably not what we want. Maybe it would be better to close the policy modal first, then open the upgrade modal. You could use the Result type to handle this.
> 
> ![Screenshot 2025-03-28 at 11 11 21 AM](https://private-user-images.githubusercontent.com/31796059/427802404-ec7a6361-9cd9-4d58-a2ec-1f63abeb591b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0MTA2ODksIm5iZiI6MTc0MzQxMDM4OSwicGF0aCI6Ii8zMTc5NjA1OS80Mjc4MDI0MDQtZWM3YTYzNjEtOWNkOS00ZDU4LWEyZWMtMWY2M2FiZWI1OTFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDA4Mzk0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5Yjg5NThjZTZjOTliNTgxMjczNDhiODhkOGFlYjE1MTIzYTVkZmJkMDA0MTJmMWE4ZmFiNzE3OWZkMzAyZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sS1RKQlE7nxWnjo-dKdKN0Z-qFeyqOhov2BALV73-vU)
> 
> 4. The designs show that when breadcrumbing applies, the edit policy modal's primary button should read "Upgrade" instead of "Submit". That change is missing from this PR. I can still try to submit a policy change and I get an error message back from the server.

I've addressed points 2 and 4.

Regards point 1: All the badges are appearing to be aligned the same way. I cannot find an instance where it is as indicated in the design. It is not clear whether I should prioritize the design or what we have currently in production.

Regards point 3: That's a good solution, I couldn't think of it at the time. Thank you!

eliykat commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.html:
```diff
@@ -1,12 +1,12 @@
+        variant="primary"
       >
         {{ "planNameEnterprise" | i18n }}
       </span>
```
This should be a `<button>` element. Per the [component library docs](https://components.bitwarden.com/?path=/docs/component-library-badge--docs):

> The Badge directive can be used on a `<span>` (non clickable events), or an `<a>` or `<button>` tag for interactive events.

Then you don't need `tw-cursor-pointor`.

eliykat commented on apps/web/src/app/admin-console/organizations/policies/policy-edit.component.html:
```diff
@@ -35,6 +36,18 @@
       </button>
+      <ng-template #breadcrumbing>
+        <button
+          *ngIf="isBreadcrumbingEnabled$ | async; else breadcrumbing"
```
This `ngIf` should be unnecessary because it's in an `ng-template` - it'll only be rendered if required by the `else` syntax above.

eliykat commented on apps/web/src/app/admin-console/organizations/policies/policies.component.html:
```diff
@@ -1,4 +1,16 @@
-<app-header></app-header>
+<app-header>
+  @let organization = organization$ | async;
+  <span
+    bitBadge
+    class="tw-cursor-pointer !tw-align-middle"
+    (click)="changePlan(organization)"
+    *ngIf="isBreadcrumbingEnabled$ | async"
+    slot="title-suffix"
+    variant="primary"
+  >
+    {{ "upgrade" | i18n }}
+  </span>
```
This should also be a `<button>`.

eliykat reviewed (changes requested):
A couple of final minor changes, otherwise looks good to go. Thanks for iterating on it!
